api/types/database.go

```diff
@@ -565,13 +566,32 @@ func (d *DatabaseV3) setStaticFields() {
 	d.Version = V3
 }
 
+// validDatabaseNameRegexp filters the allowed characters in database names.
+// This is the (almost) the same regexp used to check for valid DNS 1035 labels,
+// except we allow uppercase chars.
+var validDatabaseNameRegexp = regexp.MustCompile(`^[a-zA-Z]([-a-zA-Z0-9]*[a-zA-Z0-9])?$`)
+
+// ValidateDatabaseName returns an error if a given string is not a valid
+// Database name.
+// Unlike application access proxy, database name doesn't necessarily
+// need to be a valid subdomain but use the same validation logic for the
+// simplicity and consistency, except two differences: don't restrict names to
+// 63 chars in length and allow upper case chars.
+func ValidateDatabaseName(name string) error {
+	return ValidateResourceName(validDatabaseNameRegexp, name)
+}
+
 // CheckAndSetDefaults checks and sets default values for any missing fields.
 func (d *DatabaseV3) CheckAndSetDefaults() error {
 	d.setStaticFields()
 	if err := d.Metadata.CheckAndSetDefaults(); err != nil {
 		return trace.Wrap(err)
 	}
 
+	if err := ValidateDatabaseName(d.GetName()); err != nil {
```

Comment:
This might be a breaking change for current users :cry: 
We need to warn users about it

Reply:
Not sure if this is needed. 
Correctly me if I'm wrong but preview validation was more strict and there is not risk returning an error during validation of preview db resources.  

The only things that worries me with a new validation we need to ensure that all our backends supports case sensitive key name to prevent resource collisions.

Reply:
Me too. I am just afraid of breaking someone else setup and not warning them about it.

Reply:
`types.Database` is created via either `tctl`, static config, or auto discovery, but there's the legacy discovery method (within `db_service` config section, and the new way within `discovery_service` config.

`tctl`, static config,  and `discovery_service` called `ValidateDatabase` with the stricter name checking, but the `db_service` discovery method did not. So there is a chance someone is auto discovering databases with the legacy method that don't pass this name validation. I think we should warn about it.

Practically speaking, I don't think this will be an issue for AWS since they already restrict names quite a bit, for example RDS name constraints:
> Constraints: 1 to 60 alphanumeric characters or hyphens. First character must be a letter. Can't contain two consecutive hyphens. Can't end with a hyphen.

It may be an issue for Azure, which has very loose name restrictions.

Reply:
It's also used in `DatabaseServerV3.Spec.Database` [[code](https://github.com/gravitational/teleport/blob/628d77f6bb15f9c255238200b3777b9ac5c097ed/api/types/types.pb.go#L1138)]

Those are the ones I am more interested in because you are allowed to specify a static database whose name differs and now `(*DatabaseServerV3) CheckAndSetDefaults()` will start failing if that happens.

https://github.com/gravitational/teleport/blob/35b837de8734e216c6508b28823fd498901c8b17/api/types/databaseserver.go#L234-L257

Reply:
updated the v14 breaking changes in CHANGELOG.md here: [0423b76](https://github.com/gravitational/teleport/pull/28841/commits/0423b7603a66bf8f8c1c7376a5d68516495d54f3)